File: ios/Classes/Renderer/CandleRenderer.swift
import Foundation

final class CandleRenderer: ChartLayerRenderer {
  func render(frame: ChartFrame) {
    guard frame.chartType == "candle", frame.viewport.hasVisibleRange() else { return }

    let context = frame.context
    let chartRect = frame.mainChartRect
    let bodyWidth = max(1, frame.viewport.candleBodyWidth())

    for index in frame.viewport.visibleStartIndex...frame.viewport.visibleEndIndex {
      let open = frame.candleStore.open(at: index)
      let close = frame.candleStore.close(at: index)
      let high = frame.candleStore.high(at: index)
      let low = frame.candleStore.low(at: index)

      let centerX = frame.viewport.xCenter(for: index, plotLeft: chartRect.minX)
      let openY = frame.viewport.priceToY(open, top: chartRect.minY, height: chartRect.height)
      let closeY = frame.viewport.priceToY(close, top: chartRect.minY, height: chartRect.height)
      let highY = frame.viewport.priceToY(high, top: chartRect.minY, height: chartRect.height)
      let lowY = frame.viewport.priceToY(low, top: chartRect.minY, height: chartRect.height)

      let isBull = close >= open
      let color = isBull ? frame.theme.bullColor : frame.theme.bearColor

      context.setStrokeColor(color)
      context.setLineWidth(2)
      context.move(to: CGPoint(x: centerX, y: highY))
      context.addLine(to: CGPoint(x: centerX, y: lowY))
      context.strokePath()

      let top = min(openY, closeY)
      let bottom = max(openY, closeY)
      let adjustedBottom = abs(bottom - top) < 1 ? top + 1 : bottom
      let rect = CGRect(
        x: centerX - bodyWidth / 2,
        y: top,
        width: bodyWidth,
        height: adjustedBottom - top
      )
      context.setFillColor(color)
      context.fill(rect)
    }
  }
}

File: android/src/main/kotlin/com/tradechart/plugin/renderer/CandleRenderer.kt
package com.tradechart.plugin.renderer

import android.graphics.RectF
import com.tradechart.plugin.engine.ChartFrame
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

class CandleRenderer : ChartLayerRenderer {
    override fun render(frame: ChartFrame) {
        if (frame.chartType != "candle" || !frame.viewport.hasVisibleRange()) {
            return
        }

        val chartRect = frame.mainChartRect
        val bodyWidth = max(1f, frame.viewport.candleBodyWidth())

        for (index in frame.viewport.visibleStartIndex..frame.viewport.visibleEndIndex) {
            val open = frame.candleStore.openAt(index)
            val close = frame.candleStore.closeAt(index)
            val high = frame.candleStore.highAt(index)
            val low = frame.candleStore.lowAt(index)
            val centerX = frame.viewport.xCenterForIndex(index, chartRect.left)
            val openY = frame.viewport.priceToY(open, chartRect.top, chartRect.height())
            val closeY = frame.viewport.priceToY(close, chartRect.top, chartRect.height())
            val highY = frame.viewport.priceToY(high, chartRect.top, chartRect.height())
            val lowY = frame.viewport.priceToY(low, chartRect.top, chartRect.height())

            val isBull = close >= open
            val bodyPaint = if (isBull) frame.theme.bullPaint else frame.theme.bearPaint
            val wickPaint = if (isBull) frame.theme.bullWickPaint else frame.theme.bearWickPaint

            frame.canvas.drawLine(centerX, highY, centerX, lowY, wickPaint)

            val bodyTop = min(openY, closeY)
            val bodyBottom = max(openY, closeY)
            val adjustedBottom = if (abs(bodyBottom - bodyTop) < 1f) bodyTop + 1f else bodyBottom
            frame.canvas.drawRect(
                RectF(
                    centerX - bodyWidth / 2f,
                    bodyTop,
                    centerX + bodyWidth / 2f,
                    adjustedBottom,
                ),
                bodyPaint,
            )
        }
    }
}
